README.md
[![GitHub version](https://badge.fury.io/gh/sertansenturk%2Fpredominantmelodymakam.svg)](https://badge.fury.io/gh/sertansenturk%2Fpredominantmelodymakam) [![Build Status](https://travis-ci.org/sertansenturk/predominantmelodymakam.svg?branch=master)](https://travis-ci.org/sertansenturk/predominantmelodymakam) [![Code Climate](https://codeclimate.com/github/sertansenturk/predominantmelodymakam/badges/gpa.svg)](https://codeclimate.com/github/sertansenturk/predominantmelodymakam) [![DOI](https://zenodo.org/badge/21104/sertansenturk/predominantmelodymakam.svg)](https://zenodo.org/badge/latestdoi/21104/sertansenturk/predominantmelodymakam) [![License: AGPL v3](https://img.shields.io/badge/License-AGPL%20v3-ff69b4.svg)](http://www.gnu.org/licenses/agpl-3.0)

# predominantmelodymakam
Predominant melody extraction for makam music

This repository hosts the Python (2 & 3) implementation of the predominant melody extraction method proposed for makam music in:

> Atlı, H. S., Uyar, B., Şentürk, S., Bozkurt, B., and Serra, X. (2014). Audio feature extraction for exploring Turkish makam music. In Proceedings of 3rd International Conference on Audio Technologies for Music and Media, pages 142–153, Ankara, Turkey.

If you are using this extractor please cite the above paper. 

The extractor is based on the methodology by [2]. The contour selection step in [2] is trained on the specific characteristics of Western pop music and jazz. We remove this step and use a simplified contour selection step by selecting the longest contour at each time instance and discarding all other contours. Then we use the method proposed in [3] to remove the erroneous pitch estimations and correct octave errors.
    
Usage
=======
```python
from predominantmelodymakam.predominantmelodymakam import PredominantMelodyMakam
extractor = PredominantMelodyMakam()
results = extractor.run(audiofile)
```

Please refer to [demo.ipynb](demo.ipynb) for an interactive demo.

Installation
============

If you want to install the repository, it is recommended to install the package and dependencies into a virtualenv. In the terminal, do the following:

    virtualenv env
    source env/bin/activate
    python setup.py install

If you want to be able to edit files and have the changes be reflected, then install the repo like this instead:

    pip install -e .
    
Now you can install the rest of the dependencies:

    pip install -r requirements

The algorithm uses several modules in Essentia. Follow the [instructions](http://essentia.upf.edu/documentation/installing.html) to install the library. 

__FOR PYTHON 2 ONLY:__ You should link the python bindings of Essentia in the virtual environment:

    ln -s /usr/local/lib/python2.7/dist-packages/essentia env/lib/python2.7/site-packages
    
To run the [demo.ipynb](https://github.com/sertansenturk/predominantmelodymakam/blob/master/demo.ipynb) Jupyter notebook, you need to install jupyter and matplotlib:

    pip install matplotlib jupyter

Authors
-------
Hasan Sercan Atlı    hsercanatli@gmail.com  
Sertan Senturk        contact@sertansenturk.com

License
------
The source code hosted in this repository is licensed under [Affero GPL version 3](https://www.gnu.org/licenses/agpl-3.0.en.html). 

The extracted predominant melody and the figure in this repository are licensed under [Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License](http://creativecommons.org/licenses/by-nc-sa/4.0/).

Acknowledgements
------
We would like to thank Dr. Robert Garfias for allowing us to use [his makam music collection](https://eee.uci.edu/programs/rgarfias/films.html) in our research (in this repository the recording with MBID: [d2731692-626d-4a6d-9b67-a70c9e7b9745](http://musicbrainz.org/recording/d2731692-626d-4a6d-9b67-a70c9e7b9745)).

References
-------
[1] _Atlı, H. S., Uyar, B., Şentürk, S., Bozkurt, B., and Serra, X. (2014). Audio feature extraction for exploring Turkish makam music. In Proceedings of 3rd International Conference on Audio Technologies for Music and Media, Ankara, Turkey._   
[2] _Salamon, J., Gómez, E. (2012). "Melody Extraction from Polyphonic Music Signals using Pitch Contour Characteristics", IEEE Transactions on Audio, Speech and Language Processing, 20(6):1759-1770_   
[3] _Bozkurt, B. (2008). "An Automatic Pitch Analysis Method for Turkish Maqam Music," Journal of New Music Research. 37(1):1-13._


Changelog.md
#### predominantmelodymakam v1.3.0
 - Added Python 3 support (contribution by Oriol Romani: https://github.com/sertansenturk/predominantmelodymakam/pull/10)

#### predominantmelodymakam v1.2.2
 - Corrected the name of the 'pitchfilter' package in the requirements file.
 - Fixed a bug where the pitch contour selection step fails, when the audio input is very short.

#### predominantmelodymakam v1.2.1
 - Fixed a bug where the extract() method was not returning anything

#### predominantmelodymakam v1.2
 - Added the [pitchfilter](https://github.com/hsercanatli/pitchfilter) repository as a fallback if the Essentia PitchFilter does not work due to Essentia version problem (e.g. the official version for Ubuntu 14.04).
 - Added flake8 checking to Travis CI

#### predominantmelodymakam v1.1
 - Refactored the code to comply with PEP8

#### predominantmelodymakam v1.0
 - First public release


predominantmelodymakam/PredominantMelodyMakam.py
# -*- coding: utf-8 -*-
# Copyright 2014, 2015 Music Technology Group - Universitat Pompeu Fabra
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY ARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero
# General Public License v3.0 for more details.
#
# You should have received a copy of the GNU Affero General Public License v3.0
# along with this program. If not, see http://www.gnu.org/licenses/
#
# If you are using this extractor please cite the following paper:
#
# Atlı, H. S., Uyar, B., Şentürk, S., Bozkurt, B., and Serra, X. (2014). Audio
# feature extraction for exploring Turkish makam music. In Proceedings of 3rd
# International Conference on Audio Technologies for Music and Media, Ankara,
# Turkey.

from essentia import Pool
from essentia import array as e_array
import essentia.standard as estd
from math import ceil
import numpy as np


class PredominantMelodyMakam(object):
    def __init__(self, hop_size=128, frame_size=2048, bin_resolution=1,
                 min_frequency=55, max_frequency=1760, magnitude_threshold=0,
                 peak_distribution_threshold=1.4, filter_pitch=True,
                 confidence_threshold=36, min_chunk_size=50):

        self.hop_size = hop_size  # default hopSize of PredominantMelody
        self.frame_size = frame_size  # default frameSize of PredominantMelody
        self.bin_resolution = bin_resolution  # ~1/3 Hc; recommended for makams
        self.min_frequency = min_frequency  # default: minimum of
        # PitchSalienceFunction
        self.max_frequency = max_frequency  # default: maximum of
        # PitchSalienceFunction
        self.magnitude_threshold = magnitude_threshold  # default of
        # SpectralPeaks; 0 dB?
        self.peak_distribution_threshold = peak_distribution_threshold
        # default in PitchContours is 0.9; we need higher in makams 1.4
        self.filter_pitch = filter_pitch  # call PitchFilter
        self.confidence_threshold = confidence_threshold  # default
        # confidenceThreshold for pitchFilter
        self.min_chunk_size = min_chunk_size  # number of minimum allowed
        # samples of a chunk in PitchFilter; ~145 ms with
        # 128 sample hopSize & 44100 Fs

        self.sample_rate = 44100

    def get_settings(self):
        from essentia import __version__ as essentia_version
        citation = u"Atlı, H. S., Uyar, B., Şentürk, S., Bozkurt, B., " \
                   u"and Serra, X. (2014). Audio feature extraction for " \
                   u"exploring Turkish makam music. In Proceedings of 3rd " \
                   u"International Conference on Audio Technologies for " \
                   u"Music and Media, Ankara, Turkey."

        return {'hopSize': self.hop_size, 'frameSize': self.frame_size,
                'pitchUnit': 'Hz', 'binResolution': self.bin_resolution,
                'minFrequency': self.min_frequency,
                'maxFrequency': self.max_frequency,
                'magnitudeThreshold': self.magnitude_threshold,
                'peakDistributionThreshold': self.peak_distribution_threshold,
                'filterPitch': self.filter_pitch,
                'confidenceThreshold': self.confidence_threshold,
                'sampleRate': self.sample_rate,
                'minChunkSize': self.min_chunk_size,
                'essentiaVersion': essentia_version,
                'citation': citation}

    def run(self, fname):
        # Hann window with x4 zero padding
        run_windowing = estd.Windowing(zeroPadding=3 * self.frame_size)
        run_spectrum = estd.Spectrum(size=self.frame_size * 4)

        run_spectral_peaks = estd.SpectralPeaks(
            minFrequency=self.min_frequency, maxFrequency=self.max_frequency,
            magnitudeThreshold=self.magnitude_threshold,
            sampleRate=self.sample_rate, orderBy='magnitude')

        # convert unit to cents, PitchSalienceFunction takes 55 Hz as the
        # default reference
        run_pitch_salience_function = estd.PitchSalienceFunction(
            binResolution=self.bin_resolution)
        run_pitch_salience_function_peaks = estd.PitchSalienceFunctionPeaks(
            binResolution=self.bin_resolution, minFrequency=self.min_frequency,
            maxFrequency=self.max_frequency)
        run_pitch_contours = estd.PitchContours(
            hopSize=self.hop_size, binResolution=self.bin_resolution,
            peakDistributionThreshold=self.peak_distribution_threshold)

        pool = Pool()

        # load audio and eqLoudness
        # Note: MonoLoader resamples the audio signal to 44100 Hz by default
        audio = estd.MonoLoader(filename=fname)()
        audio = estd.EqualLoudness()(audio)

        for frame in estd.FrameGenerator(audio, frameSize=self.frame_size,
                                         hopSize=self.hop_size):
            frame = run_windowing(frame)
            spectrum = run_spectrum(frame)
            peak_frequencies, peak_magnitudes = run_spectral_peaks(spectrum)
            salience = run_pitch_salience_function(peak_frequencies,
                                                   peak_magnitudes)
            salience_peaks_bins, salience_peaks_contour_saliences = \
                run_pitch_salience_function_peaks(salience)
            if not np.size(salience_peaks_bins):
                salience_peaks_bins = np.array([0])
            if not np.size(salience_peaks_contour_saliences):
                salience_peaks_contour_saliences = np.array([0])

            pool.add('allframes_salience_peaks_bins', salience_peaks_bins)
            pool.add('allframes_salience_peaks_contourSaliences',
                     salience_peaks_contour_saliences)

        # post-processing: contour tracking
        contours_bins, contour_saliences, contours_start_times, duration = \
            run_pitch_contours(
                pool['allframes_salience_peaks_bins'],
                pool['allframes_salience_peaks_contourSaliences'])

        # run the simplified contour selection
        [pitch, pitch_salience] = self.select_contours(
            contours_bins, contour_saliences, contours_start_times, duration)

        # cent to Hz conversion
        pitch = [0. if p == 0
                 else 55. * 2. ** (self.bin_resolution * p / 1200.)
                 for p in pitch]
        pitch = e_array(pitch)
        pitch_salience = e_array(pitch_salience)

        # pitch filter
        if self.filter_pitch:
            try:
                run_pitch_filter = estd.PitchFilter(
                    confidenceThreshold=self.confidence_threshold,
                    minChunkSize=self.min_chunk_size)
                pitch = run_pitch_filter(pitch, pitch_salience)
                # generate time stamps
            except AttributeError:  # fall back to python implementation
                from pitchfilter.PitchFilter import PitchFilter
                run_pitch_filter = PitchFilter()

                # generate time stamps
                temp_tt = [s * self.hop_size / float(
                    self.sample_rate) for s in xrange(0, len(pitch))]
                temp_pitch = np.vstack((
                    temp_tt, pitch, pitch_salience)).transpose()

                temp_pitch = run_pitch_filter.run(temp_pitch)

                pitch = temp_pitch[:, 1]
                pitch_salience = temp_pitch[:, 2]

        time_stamps = [s * self.hop_size / float(self.sample_rate)
                       for s in xrange(0, len(pitch))]
        # [time pitch salience] matrix
        out = np.transpose(
            np.vstack((time_stamps, pitch.tolist(), pitch_salience.tolist())))
        out = out.tolist()

        # settings
        settings = self.get_settings()
        settings.update({'source': fname})

        return {'pitch': out, 'settings': settings}

    def select_contours(self, pitch_contours, contour_saliences, start_times,
                        duration):
        sample_rate = self.sample_rate

        hop_size = self.hop_size

        # number in samples in the audio
        num_samples = int(ceil((duration * sample_rate) / hop_size))

        # Start points of the contours in samples
        start_samples = [
            int(round(start_times[i] * sample_rate / float(hop_size)))
            for i in xrange(0, len(start_times))]

        pitch_contours_no_overlap = []
        start_samples_no_overlap = []
        contour_saliences_no_overlap = []
        lens_no_overlap = []
        while pitch_contours:  # terminate when all the contours are checked
            # print len(pitchContours)

            # get the lengths of the pitchContours
            lens = [len(k) for k in pitch_contours]

            # find the longest pitch contour
            long_idx = lens.index(max(lens))

            # pop the lists related to the longest pitchContour and append
            # it to the new list
            pitch_contours_no_overlap.append(pitch_contours.pop(long_idx))
            contour_saliences_no_overlap.append(
                contour_saliences.pop(long_idx))
            start_samples_no_overlap.append(start_samples.pop(long_idx))
            lens_no_overlap.append(lens.pop(long_idx))

            # accumulate the filled samples
            acc_idx = range(start_samples_no_overlap[-1],
                            start_samples_no_overlap[-1] + lens_no_overlap[-1])

            # remove overlaps
            [start_samples, pitch_contours,
             contour_saliences] = self.remove_overlaps(start_samples,
                                                       pitch_contours,
                                                       contour_saliences, lens,
                                                       acc_idx)

        # accumulate pitch and salience
        pitch = np.array([0.] * num_samples)
        salience = np.array([0.] * num_samples)
        for i in xrange(0, len(pitch_contours_no_overlap)):
            start_samp = start_samples_no_overlap[i]
            end_samp = start_samples_no_overlap[i] + len(
                pitch_contours_no_overlap[i])

            try:
                pitch[start_samp:end_samp] = pitch_contours_no_overlap[i]
                salience[start_samp:end_samp] = contour_saliences_no_overlap[i]

            except ValueError:
                print "The last pitch contour exceeds the audio length. " \
                      "Trimming..."

                pitch[start_samp:] = pitch_contours_no_overlap[i][:len(
                    pitch) - start_samp]
                salience[start_samp:] = contour_saliences_no_overlap[i][:len(
                    pitch) - start_samp]

        return pitch, salience

    @staticmethod
    def remove_overlaps(start_samples, pitch_contours, contour_saliences,
                        lens, acc_idx):
        # remove overlaps
        rmv_idx = []
        for i in xrange(0, len(start_samples)):
            # print '_' + str(i)
            # create the sample index vector for the checked pitch contour
            curr_samp_idx = range(start_samples[i], start_samples[i] + lens[i])

            # get the non-overlapping samples
            curr_samp_idx_no_overlap = list(set(curr_samp_idx) -
                                            set(acc_idx))

            if curr_samp_idx_no_overlap:
                temp = min(curr_samp_idx_no_overlap)
                keep_idx = range(temp - start_samples[i],
                                 (max(curr_samp_idx_no_overlap) -
                                  start_samples[i]) + 1)

                # remove all overlapping values
                pitch_contours[i] = np.array(pitch_contours[i])[keep_idx]
                contour_saliences[i] = np.array(contour_saliences[i])[keep_idx]
                # update the startSample
                start_samples[i] = temp
            else:  # totally overlapping
                rmv_idx.append(i)

        # remove totally overlapping pitch contours
        rmv_idx = sorted(rmv_idx, reverse=True)
        for r in rmv_idx:
            pitch_contours.pop(r)
            contour_saliences.pop(r)
            start_samples.pop(r)

        return start_samples, pitch_contours, contour_saliences


predominantmelodymakam/predominantmelodymakam.py
# -*- coding: utf-8 -*-
# Copyright 2014 - 2017 Music Technology Group - Universitat Pompeu Fabra
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY ARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero
# General Public License v3.0 for more details.
#
# You should have received a copy of the GNU Affero General Public License v3.0
# along with this program. If not, see http://www.gnu.org/licenses/
#
# If you are using this extractor please cite the following paper:
#
# Atlı, H. S., Uyar, B., Şentürk, S., Bozkurt, B., and Serra, X. (2014). Audio
# feature extraction for exploring Turkish makam music. In Proceedings of 3rd
# International Conference on Audio Technologies for Music and Media, Ankara,
# Turkey.

from essentia import Pool
from essentia import array as e_array
import essentia.standard as estd
from math import ceil
import numpy as np
import warnings


class PredominantMelodyMakam(object):
    def __init__(self, hop_size=128, frame_size=2048, bin_resolution=1.0,
                 min_frequency=55, max_frequency=1760, magnitude_threshold=0,
                 peak_distribution_threshold=1.4, filter_pitch=True,
                 confidence_threshold=36, min_chunk_size=50):

        self.hop_size = hop_size  # default hopSize of PredominantMelody
        self.frame_size = frame_size  # default frameSize of PredominantMelody
        self.bin_resolution = bin_resolution  # ~1/3 Hc; recommended for makams
        self.min_frequency = min_frequency  # default: minimum of
        # PitchSalienceFunction
        self.max_frequency = max_frequency  # default: maximum of
        # PitchSalienceFunction
        self.magnitude_threshold = magnitude_threshold  # default of
        # SpectralPeaks; 0 dB?
        self.peak_distribution_threshold = peak_distribution_threshold
        # default in PitchContours is 0.9; we need higher in makams 1.4
        self.filter_pitch = filter_pitch  # call PitchFilter
        self.confidence_threshold = confidence_threshold  # default
        # confidenceThreshold for pitchFilter
        self.min_chunk_size = min_chunk_size  # number of minimum allowed
        # samples of a chunk in PitchFilter; ~145 ms with
        # 128 sample hopSize & 44100 Fs

        self.sample_rate = 44100

    def get_settings(self):
        from essentia import __version__ as essentia_version
        citation = u"Atlı, H. S., Uyar, B., Şentürk, S., Bozkurt, B., " \
                   u"and Serra, X. (2014). Audio feature extraction for " \
                   u"exploring Turkish makam music. In Proceedings of 3rd " \
                   u"International Conference on Audio Technologies for " \
                   u"Music and Media, Ankara, Turkey."

        return {'hopSize': self.hop_size, 'frameSize': self.frame_size,
                'pitchUnit': 'Hz', 'binResolution': self.bin_resolution,
                'minFrequency': self.min_frequency,
                'maxFrequency': self.max_frequency,
                'magnitudeThreshold': self.magnitude_threshold,
                'peakDistributionThreshold': self.peak_distribution_threshold,
                'filterPitch': self.filter_pitch,
                'confidenceThreshold': self.confidence_threshold,
                'sampleRate': self.sample_rate,
                'minChunkSize': self.min_chunk_size,
                'essentiaVersion': essentia_version,
                'citation': citation}

    def run(self, fname):
        # load audio and eqLoudness
        # Note: MonoLoader resamples the audio signal to 44100 Hz by default
        audio = estd.MonoLoader(filename=fname)()
        audio = estd.EqualLoudness()(audio)

        contours_bins, contours_start_times, contour_saliences, duration = \
            self._extract_pitch_contours(audio)

        # run the simplified contour selection
        [pitch, pitch_salience] = self.select_contours(
            contours_bins, contour_saliences, contours_start_times, duration)

        # cent to Hz conversion
        pitch = [0. if p == 0
                 else 55. * 2. ** (self.bin_resolution * p / 1200.)
                 for p in pitch]
        pitch = e_array(pitch)
        pitch_salience = e_array(pitch_salience)

        # pitch filter
        if self.filter_pitch:
            pitch, pitch_salience = self._post_filter_pitch(
                pitch, pitch_salience)

        # generate time stamps
        time_stamps = self._gen_time_stamps(0, len(pitch))

        # [time pitch salience] matrix
        out = np.transpose(
            np.vstack((time_stamps, pitch.tolist(), pitch_salience.tolist())))
        out = out.tolist()

        # settings
        settings = self.get_settings()
        settings.update({'source': fname})

        return {'pitch': out, 'settings': settings}

    def extract(self, fname):
        """
        Alias of self.run
        :param fname: filename
        :return: dictionary with 'pitch' and 'settings' keys
        """
        return self.run(fname)

    def _extract_pitch_contours(self, audio):
        # Hann window with x4 zero padding
        run_windowing = estd.Windowing(zeroPadding=3 * self.frame_size)
        run_spectrum = estd.Spectrum(size=self.frame_size * 4)
        run_spectral_peaks = estd.SpectralPeaks(
            minFrequency=self.min_frequency, maxFrequency=self.max_frequency,
            magnitudeThreshold=self.magnitude_threshold,
            sampleRate=self.sample_rate, orderBy='magnitude')

        # convert unit to cents, PitchSalienceFunction takes 55 Hz as the
        # default reference
        run_pitch_salience_function = estd.PitchSalienceFunction(
            binResolution=self.bin_resolution)
        run_pitch_salience_function_peaks = estd.PitchSalienceFunctionPeaks(
            binResolution=self.bin_resolution, minFrequency=self.min_frequency,
            maxFrequency=self.max_frequency)
        run_pitch_contours = estd.PitchContours(
            hopSize=self.hop_size, binResolution=self.bin_resolution,
            peakDistributionThreshold=self.peak_distribution_threshold)

        # compute frame by frame
        pool = Pool()
        for frame in estd.FrameGenerator(audio, frameSize=self.frame_size,
                                         hopSize=self.hop_size):
            frame = run_windowing(frame)
            spectrum = run_spectrum(frame)
            peak_frequencies, peak_magnitudes = run_spectral_peaks(spectrum)
            salience = run_pitch_salience_function(peak_frequencies,
                                                   peak_magnitudes)
            salience_peaks_bins, salience_peaks_contour_saliences = \
                run_pitch_salience_function_peaks(salience)
            if not np.size(salience_peaks_bins):
                salience_peaks_bins = np.array([0])
            if not np.size(salience_peaks_contour_saliences):
                salience_peaks_contour_saliences = np.array([0])

            pool.add('allframes_salience_peaks_bins', salience_peaks_bins)
            pool.add('allframes_salience_peaks_contourSaliences',
                     salience_peaks_contour_saliences)

        # post-processing: contour tracking
        contours_bins, contour_saliences, contours_start_times, duration = \
            run_pitch_contours(
                pool['allframes_salience_peaks_bins'],
                pool['allframes_salience_peaks_contourSaliences'])
        return contours_bins, contours_start_times, contour_saliences, duration

    def _post_filter_pitch(self, pitch, pitch_salience):
        try:
            run_pitch_filter = estd.PitchFilter(
                confidenceThreshold=self.confidence_threshold,
                minChunkSize=self.min_chunk_size)
            pitch = run_pitch_filter(pitch, pitch_salience)

        except AttributeError:  # fall back to python implementation
            from pitchfilter.pitchfilter import PitchFilter
            run_pitch_filter = PitchFilter()

            # generate time stamps
            time_stamps = self._gen_time_stamps(0, len(pitch))

            temp_pitch = np.vstack((
                time_stamps, pitch, pitch_salience)).transpose()

            temp_pitch = run_pitch_filter.run(temp_pitch)

            pitch = temp_pitch[:, 1]
            pitch_salience = temp_pitch[:, 2]

        return pitch, pitch_salience

    def _gen_time_stamps(self, start_samp, end_samp):
        time_stamps = [s * self.hop_size / float(
            self.sample_rate) for s in range(start_samp, end_samp)]
        return time_stamps

    def select_contours(self, pitch_contours, contour_saliences, start_times,
                        duration):
        sample_rate = self.sample_rate

        hop_size = self.hop_size

        # number in samples in the audio
        num_samples = int(ceil((duration * sample_rate) / hop_size))

        # Start points of the contours in samples
        start_samples = [
            int(round(start_times[i] * sample_rate / float(hop_size)))
            for i in range(0, len(start_times))]

        pitch_contours_no_overlap = []
        start_samples_no_overlap = []
        contour_saliences_no_overlap = []
        lens_no_overlap = []
        try:
            # the pitch contours is a list of numpy arrays, parse them starting
            # with the longest contour
            while pitch_contours:  # terminate when all the contours are
                # checked
                # print len(pitchContours)

                # get the lengths of the pitchContours
                lens = [len(k) for k in pitch_contours]

                # find the longest pitch contour
                long_idx = lens.index(max(lens))

                # pop the lists related to the longest pitchContour and append
                # it to the new list
                pitch_contours_no_overlap.append(pitch_contours.pop(long_idx))
                contour_saliences_no_overlap.append(
                    contour_saliences.pop(long_idx))
                start_samples_no_overlap.append(start_samples.pop(long_idx))
                lens_no_overlap.append(lens.pop(long_idx))

                # accumulate the filled samples
                acc_idx = range(start_samples_no_overlap[-1],
                                start_samples_no_overlap[-1] +
                                lens_no_overlap[-1])

                # remove overlaps
                [start_samples, pitch_contours, contour_saliences] = \
                    self._remove_overlaps(start_samples, pitch_contours,
                                          contour_saliences, lens, acc_idx)
        except ValueError:
            # if the audio input is very short such that Essentia returns a
            # single contour as a numpy array (of length 1) of numpy array
            # (of length 1). In this case the while loop fails directly
            # as it tries to check all the truth value of an all pitch values,
            # instead of checking whether the list is empty or not.
            # Here we handle the error in a Pythonic way by simply breaking the
            # loop and assigning the inputs to outputs since a single contour
            # means nothing to filter
            pitch_contours_no_overlap = pitch_contours
            contour_saliences_no_overlap = contour_saliences
            start_samples_no_overlap = start_samples

        pitch, salience = self._join_contours(pitch_contours_no_overlap,
                                              contour_saliences_no_overlap,
                                              start_samples_no_overlap,
                                              num_samples)

        return pitch, salience

    @staticmethod
    def _join_contours(pitch_contours_no_overlap, contour_saliences_no_overlap,
                       start_samples_no_overlap, num_samples):
        # accumulate pitch and salience
        pitch = np.array([0.] * num_samples)
        salience = np.array([0.] * num_samples)
        for i in range(0, len(pitch_contours_no_overlap)):
            start_samp = start_samples_no_overlap[i]
            end_samp = start_samples_no_overlap[i] + len(
                pitch_contours_no_overlap[i])

            try:
                pitch[start_samp:end_samp] = pitch_contours_no_overlap[i]
                salience[start_samp:end_samp] = contour_saliences_no_overlap[i]
            except ValueError:
                warnings.warn("The last pitch contour exceeds the audio "
                              "length. Trimming...")

                pitch[start_samp:] = pitch_contours_no_overlap[i][:len(
                    pitch) - start_samp]
                salience[start_samp:] = contour_saliences_no_overlap[i][:len(
                    salience) - start_samp]
        return pitch, salience

    @staticmethod
    def _remove_overlaps(start_samples, pitch_contours, contour_saliences,
                         lens, acc_idx):
        # remove overlaps
        rmv_idx = []
        for i in range(0, len(start_samples)):
            # print '_' + str(i)
            # create the sample index vector for the checked pitch contour
            curr_samp_idx = range(start_samples[i], start_samples[i] + lens[i])

            # get the non-overlapping samples
            curr_samp_idx_no_overlap = list(set(curr_samp_idx) -
                                            set(acc_idx))

            if curr_samp_idx_no_overlap:
                temp = min(curr_samp_idx_no_overlap)
                keep_idx = range(temp - start_samples[i],
                                 (max(curr_samp_idx_no_overlap) -
                                  start_samples[i]) + 1)

                # remove all overlapping values
                pitch_contours[i] = np.array(pitch_contours[i])[keep_idx]
                contour_saliences[i] = np.array(contour_saliences[i])[keep_idx]
                # update the startSample
                start_samples[i] = temp
            else:  # totally overlapping
                rmv_idx.append(i)

        # remove totally overlapping pitch contours
        rmv_idx = sorted(rmv_idx, reverse=True)
        for r in rmv_idx:
            pitch_contours.pop(r)
            contour_saliences.pop(r)
            start_samples.pop(r)

        return start_samples, pitch_contours, contour_saliences
